README.md
# Optimize WP SDK

The WP optimize package for [Karolína Vyskočilová's](https://kybernaut.cz) WordPress projects. Feel free to use it, if it helps.

## Install

Via Composer

``` bash
$ composer require vyskoczilova/optimize-wp-sdk
```

## Usage

``` php
// Load Optimize WP
$composer_autoload = __DIR__ . '/vendor/autoload.php';
if ( file_exists( $composer_autoload ) ) {
	require_once $composer_autoload;
	new Vyskoczilova\OptimizeWP();
}
```

## Contributing

Please see [CONTRIBUTING](https://github.com/vyskoczilova/optimize-wp-sdk/blob/master/CONTRIBUTING.md) for details.

## Credits

* [Karolína Vyskočilová](https://github.com/vyskoczilova)
* [All Contributors](https://github.com/vyskoczilova/optimize-wp-sdk/contributors)
* [SVG-sanitizer library](https://github.com/darylldoyle/svg-sanitizer)

## License

The MIT License (MIT). Please see [License File](LICENSE.md) for more information.


CHANGELOG.md
# Changelog

All Notable changes to `optimize-wp-sdk` will be documented in this file

## 0.1.0 (2020-03-10)
* Add SVG-sanitize to sanitize the SVG files

## 0.0.2 (2020-03-09)

* Initial release.

src/SVGSupport.php
<?php

namespace Vyskoczilova;
use enshrined\svgSanitize;

class SVGSupport
{

    /**
     * The sanitizer
     *
     * @var \enshrined\svgSanitize\Sanitizer
     */
    protected $sanitizer;

    /**
     * Create a new OptimizeWP Instance
     */
    public function __construct()
    {
        $this->sanitizer = new \svgSanitize\Sanitizer();
        $this->sanitizer->minify( true );

        add_filter( 'upload_mimes', array( $this, 'svg_upload_mimes') );
        add_filter( 'wp_check_filetype_and_ext', array( $this, 'svgs_disable_real_mime_check'), 10, 4 );
        add_filter( 'wp_prepare_attachment_for_js', array( $this, 'svgs_response_for_svg'), 10, 3 );  
        add_filter( 'wp_handle_upload_prefilter', array( $this, 'check_for_svg' ) );      
    }

    /**
     * Check if the file is an SVG, if so handle appropriately
     *
     * @param $file
     *
     * @return mixed
     */
    public function check_for_svg( $file ) {

        if ( $file['type'] === 'image/svg+xml' ) {
            if ( ! $this->sanitize( $file['tmp_name'] ) ) {
                $file['error'] = __( "Sorry, this file couldn't be sanitized so for security reasons wasn't uploaded",
                    'safe-svg' );
            }
        }

        return $file;
    }

    /**
     * Sanitize the SVG
     *
     * @param $file
     *
     * @return bool|int
     */
    protected function sanitize( $file ) {
        $dirty = file_get_contents( $file );

        // Is the SVG gzipped? If so we try and decode the string
        if ( $is_zipped = $this->is_gzipped( $dirty ) ) {
            $dirty = gzdecode( $dirty );

            // If decoding fails, bail as we're not secure
            if ( $dirty === false ) {
                return false;
            }
        }

        /**
         * Load extra filters to allow devs to access the safe tags and attrs by themselves.
         */
        $this->sanitizer->setAllowedTags( new safe_svg_tags() );
        $this->sanitizer->setAllowedAttrs( new safe_svg_attributes() );

        $clean = $this->sanitizer->sanitize( $dirty );

        if ( $clean === false ) {
            return false;
        }

        // If we were gzipped, we need to re-zip
        if ( $is_zipped ) {
            $clean = gzencode( $clean );
        }

        file_put_contents( $file, $clean );

        return true;
    }

    /**
     * Check if the contents are gzipped
     *
     * @see http://www.gzip.org/zlib/rfc-gzip.html#member-format
     *
     * @param $contents
     *
     * @return bool
     */
    protected function is_gzipped( $contents ) {
        if ( function_exists( 'mb_strpos' ) ) {
            return 0 === mb_strpos( $contents, "\x1f" . "\x8b" . "\x08" );
        } else {
            return 0 === strpos( $contents, "\x1f" . "\x8b" . "\x08" );
        }
    }

    /**
     * Filters list of allowed mime types and file extensions.
     *
     * @param array $mimes Mime types keyed by the file extension regex corresponding to
     *                     those types. 'swf' and 'exe' removed from full list. 'htm|html' also
     *                     removed depending on '$user' capabilities.
     *
     * @return array $mimes
     */
    public function svg_upload_mimes( $mimes = array() ) {
        $mimes['svg']  = 'image/svg+xml';
        $mimes['svgz'] = 'image/svg+xml';
        return $mimes;
    }

    /**
     * Filters the "real" file type of the given file.
     *
     * @param array  $data     File data array containing 'ext', 'type', and
     *                         'proper_filename' keys.
     * @param string $file     Full path to the file.
     * @param string $filename The name of the file (may differ from $file due to
     *                         $file being in a tmp directory).
     * @param array  $mimes    Key is the file extension with value as the mime type.
     *
     * @return array
     */
    public function svgs_disable_real_mime_check( $data, $file, $filename, $mimes ) {
        $wp_filetype = wp_check_filetype( $filename, $mimes );

        $ext             = $wp_filetype['ext'];
        $type            = $wp_filetype['type'];
        $proper_filename = $data['proper_filename'];

        return compact( 'ext', 'type', 'proper_filename' );
    }

    /**
     * Filters the attachment data prepared for JavaScript.
     * Base on /wp-includes/media.php
     *
     * @param array          $response   Array of prepared attachment data.
     * @param integer|object $attachment Attachment ID or object.
     * @param array          $meta       Array of attachment meta data.
     *
     * @return mixed $response
     */
    public function svgs_response_for_svg( $response, $attachment, $meta ) {
        if ( 'image/svg+xml' === $response['mime'] && empty( $response['sizes'] ) ) {
            $svg_path = get_attached_file( $attachment->ID );
            if ( ! file_exists( $svg_path ) ) {
                // If SVG is external, use the URL instead of the path.
                $svg_path = $response['url'];
            }
            $dimensions        = $this->svgs_get_dimensions( $svg_path );
            $response['sizes'] = array(
                'full' => array(
                    'url'         => $response['url'],
                    'width'       => $dimensions->width,
                    'height'      => $dimensions->height,
                    'orientation' => $dimensions->width > $dimensions->height ? 'landscape' : 'portrait',
                ),
            );
        }

        return $response;
    }

    /**
     * Get dimension svg file
     *
     * @param string $svg Path of svg.
     * @return object width and height.
     */
    private function svgs_get_dimensions( $svg ) {
        $svg = simplexml_load_file( $svg );
        if ( false === $svg ) {
            $width  = '0';
            $height = '0';
        } else {
            $attributes = $svg->attributes();
            $width      = (string) $attributes->width;
            $height     = (string) $attributes->height;
        }

        return (object) array(
            'width'  => $width,
            'height' => $height,
        );
    }

}


src/OptimizeWPClass.php
<?php

namespace Vyskoczilova\OptimizeWP;

class OptimizeWPClass
{

    /**
     * Create a new OptimizeWP Instance
     */
    public function __construct()
    {
        add_action('wp_head', array($this,'kbnt_preload_js_assets'), 1);
        add_action( 'wp_default_scripts', 'kbnt_theme_remove_jquery_migrate' );
        add_action( 'init', 'kbnt_disable_emojis' );
        add_action( 'init', 'kbnt_disable_embeds_code_init', 9999 );
        add_action( 'wp_print_styles', 'kbnt_theme_deregister_styles', 100 );

        add_filter('style_loader_tag', 'kbnt_remove_type_attr', 999, 2);
        add_filter('script_loader_tag', 'kbnt_remove_type_attr', 999, 2);
        add_filter( 'get_custom_logo', 'kbnt_remove_itemprop' );

        // Clean WP head
        remove_action( 'wp_head', 'feed_links_extra', 3 ); 
        remove_action( 'wp_head', 'feed_links', 2 ); 
        remove_action( 'wp_head', 'rsd_link'); 
        remove_action( 'wp_head', 'wlwmanifest_link'); 
        remove_action( 'wp_head', 'wp_generator'); 
        remove_action( 'wp_head', 'adjacent_posts_rel_link_wp_head', 10, 0 );

        // Handle SVG
        new SVGSupportClass();
        
    }

    /**
     * Preloading JavaScript Assets in WordPress
     * https://macarthur.me/posts/preloading-javascript-in-wordpress
     */
    public function kbnt_preload_js_assets() {
        global $wp_scripts;

        foreach($wp_scripts->queue as $handle) {
            $script = $wp_scripts->registered[$handle];

            //-- Weird way to check if script is being enqueued in the footer.
            if( isset( $script->extra['group'] ) && $script->extra['group'] === 1) {

                //-- If version is set, append to end of source.
                $source = $script->src . ($script->ver ? "?ver={$script->ver}" : "");

                //-- Spit out the tag.
                echo "<link rel='preload' href='{$source}' as='script'/>\n";
            }
        }
    }

    /**
     * Remove type attribute from style and scripts
     * W3 WARNING:
     *  The type attribute for the style element is not needed and should be omitted.
     *  The type attribute is unnecessary for JavaScript resources.
     */ 
    public function kbnt_remove_type_attr($tag, $handle) {
        return preg_replace( " /type=['\"]text\/(javascript|css)['\"]/", '', $tag );
    }

    /**
     * W3 ERROR: The itemprop attribute was specified, but the element is not a property of any item.
     */
    public function kbnt_remove_itemprop() {
        $custom_logo_id = get_theme_mod( 'custom_logo' );
        $html = sprintf( '<a href="%1$s" class="custom-logo-link" rel="home">%2$s</a>',
                esc_url( home_url( '/' ) ),
                wp_get_attachment_image( $custom_logo_id, 'full', false, array(
                    'class'    => 'custom-logo',
                ) )
            );
        return $html;   
    }

    /**
     * Remove jQuery migrate
     *
     * @param array $scripts
     * @return array
     */
    public function kbnt_theme_remove_jquery_migrate( $scripts ) {
        if ( ! is_admin() && isset( $scripts->registered['jquery'] ) ) {
            $script = $scripts->registered['jquery'];
            
            if ( $script->deps ) { // Check whether the script has any dependencies
                $script->deps = array_diff( $script->deps, array( 'jquery-migrate' ) );
            }
        }
    }

    /**
     * Disable emojis
     *
     * @return void
     */
    public function kbnt_disable_emojis() {
        remove_action( 'wp_head', 'print_emoji_detection_script', 7 );
        remove_action( 'admin_print_scripts', 'print_emoji_detection_script' );
        remove_action( 'wp_print_styles', 'print_emoji_styles' );
        remove_action( 'admin_print_styles', 'print_emoji_styles' );	
        remove_filter( 'the_content_feed', 'wp_staticize_emoji' );
        remove_filter( 'comment_text_rss', 'wp_staticize_emoji' );	
        remove_filter( 'wp_mail', 'wp_staticize_emoji_for_email' );
        add_filter( 'tiny_mce_plugins', 'kbnt_disable_emojis_tinymce' );
        add_filter( 'wp_resource_hints', 'kbnt_disable_emojis_remove_dns_prefetch', 10, 2 );
    }

    /**
     * Disable WP embeds
     * See https://kinsta.com/knowledgebase/disable-embeds-wordpress/.
     *
     * @return void
     */
    public function kbnt_disable_embeds_code_init() {

        // Remove the REST API endpoint.
        remove_action( 'rest_api_init', 'wp_oembed_register_route' );
    
        // Turn off oEmbed auto discovery.
        add_filter( 'embed_oembed_discover', '__return_false' );
    
        // Don't filter oEmbed results.
        remove_filter( 'oembed_dataparse', 'wp_filter_oembed_result', 10 );
    
        // Remove oEmbed discovery links.
        remove_action( 'wp_head', 'wp_oembed_add_discovery_links' );
    
        // Remove oEmbed-specific JavaScript from the front-end and back-end.
        remove_action( 'wp_head', 'wp_oembed_add_host_js' );
        add_filter( 'tiny_mce_plugins', array( $this, 'kbnt_disable_embeds_tiny_mce_plugin') );
    
        // Remove all embeds rewrite rules.
        add_filter( 'rewrite_rules_array', array( $this, 'kbnt_disable_embeds_rewrites') );
    
        // Remove filter of the oEmbed result before any HTTP requests are made.
        remove_filter( 'pre_oembed_result', 'wp_filter_pre_oembed_result', 10 );
    }

    public function kbnt_disable_embeds_tiny_mce_plugin($plugins) {
        return array_diff($plugins, array('wpembed'));
    }
    
    public function kbnt_disable_embeds_rewrites($rules) {
        foreach($rules as $rule => $rewrite) {
            if(false !== strpos($rewrite, 'embed=true')) {
                unset($rules[$rule]);
            }
        }
        return $rules;
    }

    /**
     * Remove dashicons in frontend for unauthenticated users
     *
     * @return void
     */
    public function kbnt_theme_deregister_styles() { 
        if ( ! is_user_logged_in() ) {
            wp_dequeue_style( 'dashicons-css' );
            wp_deregister_style( 'dashicons-css' );
        }
    }

}


src/SVGSupportClass.php
<?php

namespace Vyskoczilova\OptimizeWP;

class SVGSupportClass
{

    /**
     * Create a new OptimizeWP Instance
     */
    public function __construct()
    {
        add_filter( 'upload_mimes', 'svg_upload_mimes' );
        add_filter( 'wp_check_filetype_and_ext', 'svgs_disable_real_mime_check', 10, 4 );
        add_filter( 'wp_prepare_attachment_for_js', 'svgs_response_for_svg', 10, 3 );        
    }

    /**
     * Filters list of allowed mime types and file extensions.
     *
     * @param array $mimes Mime types keyed by the file extension regex corresponding to
     *                     those types. 'swf' and 'exe' removed from full list. 'htm|html' also
     *                     removed depending on '$user' capabilities.
     *
     * @return array $mimes
     */
    public function svg_upload_mimes( $mimes = array() ) {
        $mimes['svg']  = 'image/svg+xml';
        $mimes['svgz'] = 'image/svg+xml';
        return $mimes;
    }

    /**
     * Filters the "real" file type of the given file.
     *
     * @param array  $data     File data array containing 'ext', 'type', and
     *                         'proper_filename' keys.
     * @param string $file     Full path to the file.
     * @param string $filename The name of the file (may differ from $file due to
     *                         $file being in a tmp directory).
     * @param array  $mimes    Key is the file extension with value as the mime type.
     *
     * @return array
     */
    public function svgs_disable_real_mime_check( $data, $file, $filename, $mimes ) {
        $wp_filetype = wp_check_filetype( $filename, $mimes );

        $ext             = $wp_filetype['ext'];
        $type            = $wp_filetype['type'];
        $proper_filename = $data['proper_filename'];

        return compact( 'ext', 'type', 'proper_filename' );
    }

    /**
     * Filters the attachment data prepared for JavaScript.
     * Base on /wp-includes/media.php
     *
     * @param array          $response   Array of prepared attachment data.
     * @param integer|object $attachment Attachment ID or object.
     * @param array          $meta       Array of attachment meta data.
     *
     * @return mixed $response
     */
    public function svgs_response_for_svg( $response, $attachment, $meta ) {
        if ( 'image/svg+xml' === $response['mime'] && empty( $response['sizes'] ) ) {
            $svg_path = get_attached_file( $attachment->ID );
            if ( ! file_exists( $svg_path ) ) {
                // If SVG is external, use the URL instead of the path.
                $svg_path = $response['url'];
            }
            $dimensions        = $this->svgs_get_dimensions( $svg_path );
            $response['sizes'] = array(
                'full' => array(
                    'url'         => $response['url'],
                    'width'       => $dimensions->width,
                    'height'      => $dimensions->height,
                    'orientation' => $dimensions->width > $dimensions->height ? 'landscape' : 'portrait',
                ),
            );
        }

        return $response;
    }

    /**
     * Get dimension svg file
     *
     * @param string $svg Path of svg.
     * @return object width and height.
     */
    private function svgs_get_dimensions( $svg ) {
        $svg = simplexml_load_file( $svg );
        if ( false === $svg ) {
            $width  = '0';
            $height = '0';
        } else {
            $attributes = $svg->attributes();
            $width      = (string) $attributes->width;
            $height     = (string) $attributes->height;
        }

        return (object) array(
            'width'  => $width,
            'height' => $height,
        );
    }

}


src/OptimizeWP.php
<?php

namespace Vyskoczilova;

class OptimizeWP
{

    /**
     * Create a new OptimizeWP Instance
     */
    public function __construct()
    {
        add_action('wp_head', array($this,'preload_js_assets'), 1);
        add_action( 'wp_default_scripts',  array($this,'theme_remove_jquery_migrate' ));
        add_action( 'init',  array($this,'disable_emojis' ));
        add_action( 'init',  array($this,'disable_embeds_code_init'), 9999 );
        add_action( 'wp_print_styles',  array($this,'theme_deregister_styles'), 100 );

        add_filter('style_loader_tag', array($this,'remove_type_attr'), 999, 2);
        add_filter('script_loader_tag', array($this,'remove_type_attr'), 999, 2);
        add_filter( 'get_custom_logo',  array($this,'remove_itemprop' ));

        // Clean WP head
        remove_action( 'wp_head', 'feed_links_extra', 3 ); 
        remove_action( 'wp_head', 'feed_links', 2 ); 
        remove_action( 'wp_head', 'rsd_link'); 
        remove_action( 'wp_head', 'wlwmanifest_link'); 
        remove_action( 'wp_head', 'wp_generator'); 
        remove_action( 'wp_head', 'adjacent_posts_rel_link_wp_head', 10, 0 );

        // Handle SVG
        new SVGSupport();
        
    }

    /**
     * Preloading JavaScript Assets in WordPress
     * https://macarthur.me/posts/preloading-javascript-in-wordpress
     */
    public function preload_js_assets() {
        global $wp_scripts;

        foreach($wp_scripts->queue as $handle) {
            $script = $wp_scripts->registered[$handle];

            //-- Weird way to check if script is being enqueued in the footer.
            if( isset( $script->extra['group'] ) && $script->extra['group'] === 1) {

                //-- If version is set, append to end of source.
                $source = $script->src . ($script->ver ? "?ver={$script->ver}" : "");

                //-- Spit out the tag.
                echo "<link rel='preload' href='{$source}' as='script'/>\n";
            }
        }
    }

    /**
     * Remove type attribute from style and scripts
     * W3 WARNING:
     *  The type attribute for the style element is not needed and should be omitted.
     *  The type attribute is unnecessary for JavaScript resources.
     */ 
    public function remove_type_attr($tag, $handle) {
        return preg_replace( " /type=['\"]text\/(javascript|css)['\"]/", '', $tag );
    }

    /**
     * W3 ERROR: The itemprop attribute was specified, but the element is not a property of any item.
     */
    public function remove_itemprop() {
        $custom_logo_id = get_theme_mod( 'custom_logo' );
        $html = sprintf( '<a href="%1$s" class="custom-logo-link" rel="home">%2$s</a>',
                esc_url( home_url( '/' ) ),
                wp_get_attachment_image( $custom_logo_id, 'full', false, array(
                    'class'    => 'custom-logo',
                ) )
            );
        return $html;   
    }

    /**
     * Remove jQuery migrate
     *
     * @param array $scripts
     * @return array
     */
    public function theme_remove_jquery_migrate( $scripts ) {
        if ( ! is_admin() && isset( $scripts->registered['jquery'] ) ) {
            $script = $scripts->registered['jquery'];
            
            if ( $script->deps ) { // Check whether the script has any dependencies
                $script->deps = array_diff( $script->deps, array( 'jquery-migrate' ) );
            }
        }
    }

    /**
     * Disable emojis
     *
     * @return void
     */
    public function disable_emojis() {
        remove_action( 'wp_head', 'print_emoji_detection_script', 7 );
        remove_action( 'admin_print_scripts', 'print_emoji_detection_script' );
        remove_action( 'wp_print_styles', 'print_emoji_styles' );
        remove_action( 'admin_print_styles', 'print_emoji_styles' );	
        remove_filter( 'the_content_feed', 'wp_staticize_emoji' );
        remove_filter( 'comment_text_rss', 'wp_staticize_emoji' );	
        remove_filter( 'wp_mail', 'wp_staticize_emoji_for_email' );
        add_filter( 'tiny_mce_plugins', array($this,'disable_emojis_tinymce') );
        add_filter( 'wp_resource_hints', array($this,'disable_emojis_remove_dns_prefetch'), 10, 2 );
    }

    
    /**
     * Filter function used to remove the tinymce emoji plugin.
     * 
     * @param    array  $plugins  
     * @return   array              Difference betwen the two arrays
     */
    function disable_emojis_tinymce( $plugins ) {
        if ( is_array( $plugins ) ) {
            return array_diff( $plugins, array( 'wpemoji' ) );
        }
        
        return array();

    }

    /**
     * Disable WP embeds
     * See https://kinsta.com/knowledgebase/disable-embeds-wordpress/.
     *
     * @return void
     */
    public function disable_embeds_code_init() {

        // Remove the REST API endpoint.
        remove_action( 'rest_api_init', 'wp_oembed_register_route' );
    
        // Turn off oEmbed auto discovery.
        add_filter( 'embed_oembed_discover', '__return_false' );
    
        // Don't filter oEmbed results.
        remove_filter( 'oembed_dataparse', 'wp_filter_oembed_result', 10 );
    
        // Remove oEmbed discovery links.
        remove_action( 'wp_head', 'wp_oembed_add_discovery_links' );
    
        // Remove oEmbed-specific JavaScript from the front-end and back-end.
        remove_action( 'wp_head', 'wp_oembed_add_host_js' );
        add_filter( 'tiny_mce_plugins', array( $this, 'disable_embeds_tiny_mce_plugin') );
    
        // Remove all embeds rewrite rules.
        add_filter( 'rewrite_rules_array', array( $this, 'disable_embeds_rewrites') );
    
        // Remove filter of the oEmbed result before any HTTP requests are made.
        remove_filter( 'pre_oembed_result', 'wp_filter_pre_oembed_result', 10 );
    }

    public function disable_embeds_tiny_mce_plugin($plugins) {
        return array_diff($plugins, array('wpembed'));
    }
    
    public function disable_embeds_rewrites($rules) {
        foreach($rules as $rule => $rewrite) {
            if(false !== strpos($rewrite, 'embed=true')) {
                unset($rules[$rule]);
            }
        }
        return $rules;
    }

    /**
     * Remove dashicons in frontend for unauthenticated users
     *
     * @return void
     */
    public function theme_deregister_styles() { 
        if ( ! is_user_logged_in() ) {
            wp_dequeue_style( 'dashicons-css' );
            wp_deregister_style( 'dashicons-css' );
        }
    }

    
    /**
     * Remove emoji CDN hostname from DNS prefetching hints.
     *
     * @param  array  $urls          URLs to print for resource hints.
     * @param  string $relation_type The relation type the URLs are printed for.
     * @return array                 Difference betwen the two arrays.
     */
    function disable_emojis_remove_dns_prefetch( $urls, $relation_type ) {

        if ( 'dns-prefetch' == $relation_type ) {

            // Strip out any URLs referencing the WordPress.org emoji location
            $emoji_svg_url_bit = 'https://s.w.org/images/core/emoji/';
            foreach ( $urls as $key => $url ) {
                if ( strpos( $url, $emoji_svg_url_bit ) !== false ) {
                    unset( $urls[$key] );
                }
            }

        }

        return $urls;
    }

}
